service/history/historybuilder/history_builder_test.go

```diff
@@ -267,6 +270,126 @@ func (s *historyBuilderSuite) TestWorkflowExecutionStarted() {
 				},
 				ParentInitiatedEventId:      testParentInitiatedID,
 				ParentInitiatedEventVersion: testParentInitiatedVersion,
+				RootWorkflowExecution:       nil,
+			},
+		},
+	}, event)
+}
+
+func (s *historyBuilderSuite) TestWorkflowExecutionStarted_WithRoot() {
+	attempt := rand.Int31()
+	workflowExecutionExpirationTime := timestamppb.New(time.Unix(0, rand.Int63()))
+	continueAsNewInitiator := enumspb.ContinueAsNewInitiator(rand.Int31n(int32(len(enumspb.ContinueAsNewInitiator_name))))
+	firstWorkflowTaskBackoff := durationpb.New(time.Duration(rand.Int63()))
+
+	workflowExecutionTimeout := durationpb.New(time.Duration(rand.Int63()))
+	workflowRunTimeout := durationpb.New(time.Duration(rand.Int63()))
+	workflowTaskStartToCloseTimeout := durationpb.New(time.Duration(rand.Int63()))
+
+	resetPoints := &workflowpb.ResetPoints{}
+	prevRunID := uuid.New()
+	firstRunID := uuid.New()
+	originalRunID := uuid.New()
+
+	request := &historyservice.StartWorkflowExecutionRequest{
+		NamespaceId: testNamespaceID.String(),
+		ParentExecutionInfo: &workflowspb.ParentExecutionInfo{
+			NamespaceId: testParentNamespaceID,
+			Namespace:   testParentNamespaceName,
+			Execution: &commonpb.WorkflowExecution{
+				WorkflowId: testParentWorkflowID,
+				RunId:      testParentRunID,
+			},
+			InitiatedId:      testParentInitiatedID,
+			InitiatedVersion: testParentInitiatedVersion,
+		},
+		Attempt:                         attempt,
+		WorkflowExecutionExpirationTime: workflowExecutionExpirationTime,
+		ContinueAsNewInitiator:          continueAsNewInitiator,
+		ContinuedFailure:                testFailure,
+		LastCompletionResult:            testPayloads,
+		FirstWorkflowTaskBackoff:        firstWorkflowTaskBackoff,
+		RootExecutionInfo: &workflowspb.RootExecutionInfo{
+			Execution: &commonpb.WorkflowExecution{
+				WorkflowId: testRootWorkflowID,
+				RunId:      testRootRunID,
+			},
+		},
+
+		StartRequest: &workflowservice.StartWorkflowExecutionRequest{
+			Namespace:                testNamespaceName.String(),
+			WorkflowId:               testWorkflowID,
+			WorkflowType:             testWorkflowType,
+			TaskQueue:                testTaskQueue,
+			Input:                    testPayloads,
+			WorkflowExecutionTimeout: workflowExecutionTimeout,
+			WorkflowRunTimeout:       workflowRunTimeout,
+			WorkflowTaskTimeout:      workflowTaskStartToCloseTimeout,
+			Identity:                 testIdentity,
+			RequestId:                testRequestID,
+			// WorkflowIdReusePolicy: not used for event generation
+			RetryPolicy:      testRetryPolicy,
+			CronSchedule:     testCronSchedule,
+			Memo:             testMemo,
+			SearchAttributes: testSearchAttributes,
+			Header:           testHeader,
+		},
+	}
+
+	event := s.historyBuilder.AddWorkflowExecutionStartedEvent(
+		s.now,
+		request,
+		resetPoints,
+		prevRunID,
+		firstRunID,
+		originalRunID,
+	)
+	s.Equal(event, s.flush())
+	s.Equal(&historypb.HistoryEvent{
```

Comment:
You're lucky that testify's Equal is working here; it won't if anything mucks with the internal proto-serialization state.

When comparing proto structs you should use the protorequire and protoassert ProtoEqual method